database/postgres/dal/refresh.py

```diff
@@ -0,0 +1,121 @@
+                refresh_entry.last_device = last_device
+                refresh_entry.nonce = refresh_dto.nonce
+            else:
+                refresh_entry = RefreshTokenTracking(
+                    user_id=refresh_dto.user_id,
+                    last_ip=last_ip,
+                    last_device=last_device,
+                    nonce=nonce
+                )
+                self.db_session.add(refresh_entry)
+
+            self.db_session.commit()
+            self.db_session.refresh(refresh_entry)
+            return refresh_entry.id
+        except SQLAlchemyError as e:
+            self.db_session.rollback()
+            print(f"Error in save_refresh_token: {e}")
+            return None
+
+
+    def get_refresh_token_by_id(self, refresh_id: int) -> Optional[RefreshTokenTracking]:
+        return self.db_session.query(RefreshTokenTracking).filter(RefreshTokenTracking.id == refresh_id).first()
+
+    def get_valid_refresh_token_by_user(self, user_id: int) -> Optional[RefreshTokenTracking]:
+        return (
+            self.db_session.query(RefreshTokenTracking)
+            .filter(
+                RefreshTokenTracking.user_id == user_id,
+                RefreshTokenTracking.refresh_token.has(revoked=False),  
+                TokenBlocklist.expires_at > datetime.utcnow()
+            )
+            .first()
+        )
+
+    def get_nonce_by_user_id(self, user_id: int) -> Optional[str]:
+        entry = self.db_session.query(RefreshTokenTracking).filter(RefreshTokenTracking.user_id == user_id).first()
+        return entry.nonce if entry else None
+    
+    def verify_nonce(self, user_id: int, nonce: str) -> bool:
+        token_entry = self.db_session.query(RefreshTokenTracking).filter_by(user_id=user_id, nonce=nonce).first()
+        return token_entry is not None
+    
+    def update_refresh_token(self, user_id: int, refresh_dto: refreshDTO):
+        try:
+            entry = (self.db_session.query(RefreshTokenTracking)
+                .filter(RefreshTokenTracking.user_id == user_id)
+                .with_for_update()
+                .first()
+            )
+            if entry:
+                entry.refresh_token = refresh_dto.refresh_token
+                entry.nonce = refresh_dto.nonce
+                entry.last_ip = refresh_dto.last_ip
+                entry.last_device = refresh_dto.last_device
+                self.db_session.commit()
+            else:
+                self.save_refresh_token(refresh_dto)
+        except SQLAlchemyError as e:
+            self.db_session.rollback()
+            print(f"Error in update_refresh_token: {e}")
+
+
+    def revoke_refresh_token(self, refresh_id: int) -> bool:
+        try:
+            refresh_entry = self.get_refresh_token_by_id(refresh_id)
+            if refresh_entry:
+                token_entry = self.db_session.query(TokenBlocklist).filter(TokenBlocklist.id == refresh_id).first()
+                if token_entry and token_entry.token_type == "refresh":
```

Comment:
hardcoded "refresh" string. 

Either create a table in DB (TokenTypes) and reference it's id, or make sure we have some alternative that'd prevent anyone from storing a token_type we know nothing about.